spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,492 @@
+
+# %% [markdown]
+# ### Create DataLoaders
+#
+# Next, we'll use the extracted Tensors to initialize a [`DictDataLoader`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.DictDataLoader.html) — as a quick recap, this is a Snorkel-specific class that inherits from the common PyTorch class and supports multiple data fields in the `X_dict` and labels in the `Y_dict`.
+#
+# In this task, we'd like to store the bag-of-words `bow_features` in our `X_dict`, and we have one set of labels (for now) correpsonding to the `spam_task`.
+
+# %%
+from snorkel.classification.data import DictDataset, DictDataLoader
+
+BATCH_SIZE = 32
+
+
+def create_dict_dataloader(X, Y, split, **kwargs):
+    """Create a DictDataLoader for bag-of-words features."""
+    ds = DictDataset(
+        name="spam_dataset",
+        split=split,
+        X_dict={"bow_features": torch.FloatTensor(X)},
+        Y_dict={"spam_task": torch.LongTensor(Y)},
+    )
+    return DictDataLoader(ds, **kwargs)
+
+
+dl_train = create_dict_dataloader(
+    X_train, Y_train, split="train", batch_size=BATCH_SIZE, shuffle=True
+)
+dl_valid = create_dict_dataloader(
+    X_valid, Y_valid, split="valid", batch_size=BATCH_SIZE, shuffle=False
+)
+dl_test = create_dict_dataloader(
+    X_test, Y_test, split="test", batch_size=BATCH_SIZE, shuffle=False
+)
+
+# %% [markdown]
+# We can inspect our datasets to confirm that they have the appropriate fields.
+
+# %%
+dl_valid.dataset
+
+# %% [markdown]
+# ### Define [`MultitaskClassifier`](https://snorkel.readthedocs.io/en/redux/packages/_autosummary/classification/snorkel.classification.MultitaskClassifier.html)
+#
+# We define a simple Multi-Layer Perceptron (MLP) architecture to learn from the `bow_features`.
+#
+# _Note: the following might feel like extra steps to define what is a very simple architecture (e.g. `sklearn`'s [`MLPClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html)), but this will lend us additional flexibility later in the pipeline!_
+
+# %% [markdown]
+# To start, we define a `module_pool` with all the PyTorch modules that we'll want to include in our network.
+
+# %%
+import torch.nn as nn
+
+bow_dim = X_train.shape[1]
+module_pool = nn.ModuleDict(
+    {
+        "mlp": nn.Sequential(nn.Linear(bow_dim, bow_dim), nn.ReLU()),
+        "prediction_head": nn.Linear(bow_dim, 2),
+    }
+)
+
+# %% [markdown]
+# Then, we specify the desired `task_flow` through each module.
+
+# %%
+from snorkel.classification.task import Operation
+
+task_flow = [
+    Operation(name="input_op", module_name="mlp", inputs=[("_input_", "bow_features")]),
+    Operation(name="head_op", module_name="prediction_head", inputs=[("input_op", 0)]),
+]
+
+# %% [markdown]
+# With these pieces, we're ready to define a [`Task`](https://snorkel.readthedocs.io/en/redux/source/snorkel.classification.html#module-snorkel.classification.task) in Snorkel for spam classification.
```

Comment:
not a valid link. please look at EVERY LINK and EVERY `from snorkel` import to make sure there's a corresponding link